### PR TITLE
Infra 3710 grafana

### PR DESCRIPTION
**Pull request by elainearbaugh**
On top of https://github.com/Affirm/cabot/pull/40, actual changes are in https://github.com/Affirm/cabot/pull/45/commits/3abb1ceae9d6291dc86a668170742f52623eb7bd.

- Models for Grafana Instance (grafana site) and GrafanaDataSource (links the name of a data source in Grafana to the GrafanaInstance and MetricsSourceBase objects)
- Views/forms/template for selecting a GrafanaInstance, dashboard, panel, and series
- Autofill ElasticcsearchStatusCheck form based on panel/series info
- ElasticsearchStatusCheck creation form

Tested creating checks based on some dashboards (web, prequal, charlotte, etc.), in general worked but more testing necessary. Ran into several issues with Grafana autofill values in the API response that I had to make special cases for--I'm sure there are more issues I haven't found yet. 

TODO: test cases for Grafana API parsing, handle more edge cases (and fail more gracefully) (these will be in https://github.com/Affirm/cabot/pull/46)

In future PRs: update/delete check views, refesh check based on dashboard changes, celery task for checking if a dashboard changes and emailing owners, etc.

### CODE REVIEW

File: cabot/metricsapp/api/grafana.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+import logging
+import requests
+import urlparse
+from django.core.exceptions import ValidationError
+
+
+logger = logging.getLogger(__name__)
+
+
+def _get_time_range(timestring):
```
Comment (mahendra) [inline cabot/metricsapp/api/grafana.py]: somebody wrote up a quick module for this - https://github.com/wroberts/pytimeparse

Not mandatory to use it. Just FYI

File: cabot/metricsapp/api/grafana.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+
+def get_dashboards(url, api_key):
+    """
+    Get data for all Grafana dashboards
+    :param url: url for the Grafana instance
+    :param api_key: API key for the Grafana instance
+    :return: api response containing dashboard id, title, uri, etc.
+    """
+    response = requests.get(urlparse.urljoin(url, 'api/search'), headers=get_grafana_auth_headers(api_key))
```
Comment (mahendra) [inline cabot/metricsapp/api/grafana.py]: I was wondering how we would solve this problem. An alternate and simpler way to do this would be to create a single request.Session instance.

```
session = requests.Session()
session.headers['Authorization'] = ...
```

Now, use this session everywhere. You can either pass the session as an argument or create a class with all the member functions.

Advantages of using requests sessions - connection pooling and other controls.

http://docs.python-requests.org/en/master/user/advanced/

For more examples of how we use it, please `vault.py` in salt - not linking here since this is a public repo

Comment (mahendra) [inline cabot/metricsapp/api/grafana.py]: Another advantage - in the future if we migrate to a different auth scheme, we don't have to update the invocation everywhere.

We can modify code in one place.

File: cabot/metricsapp/api/grafana.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+    if response.status_code == 200:
```
Comment (mahendra) [inline cabot/metricsapp/api/grafana.py]: can be replaced with
```
response.raise_for_status()
return response.json()
```

File: cabot/metricsapp/api/grafana.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+def get_dashboard_info(grafana_url, api_key, dashboard_uri):
+    """
+    Get data about a Grafana dashboard
+    :param url: url for the Grafana instance
+    :param api_key: API key for the Grafana instance
+    :param dashboard_uri: dashboard part of the url
+    :return: api response containing creator information and panel information
+    """
+    http_api = urlparse.urljoin(grafana_url, 'api/dashboards/db/')
```
Comment (mahendra) [inline cabot/metricsapp/api/grafana.py]: You can use requests mounting to make this easier - http://docs.python-requests.org/en/master/user/advanced/#transport-adapters

So, you can set up a requests.Session with a prefix of grafana_url. Then you need to do
```
session.get('/api/dashboards/db/')
```

Never tried it before. YMMV

File: cabot/metricsapp/api/grafana.py (hunk 1 above)
Comment (elainearbaugh) [inline cabot/metricsapp/api/grafana.py]: Oooh thanks for finding! this is way less janky than what I wrote

File: cabot/metricsapp/api/grafana.py (hunk 2 above)
Comment (elainearbaugh) [inline cabot/metricsapp/api/grafana.py]: Cool this is great! I added a [session](https://github.com/Affirm/cabot/blob/1e3065f0bccf471ea32ee92e24c6ccc11f47ce58/cabot/metricsapp/models/grafana.py#L45) property to the GrafanaInstance object (implementation is pretty similar to the client property for ElasticsearchSource) and a method for making [requests](https://github.com/Affirm/cabot/blob/1e3065f0bccf471ea32ee92e24c6ccc11f47ce58/cabot/metricsapp/models/grafana.py#L56) to this instance. Lmk if this is ok

File: cabot/metricsapp/api/grafana.py (hunk 4 above)
Comment (elainearbaugh) [inline cabot/metricsapp/api/grafana.py]: Couldn't get this to work but I added a method to the GrafanaInstance to make requests to urls starting with the grafana_url prefix instead

**Comment by elainearbaugh** [comment]
trigger build
